### PR TITLE
kafka: add how to debug KUDO kafka runbook

### PR DESCRIPTION
**Pull request by zmalik**
Signed-off-by: Zain Malik <zmalikshxil@gmail.com>

**What this PR does / why we need it**:
first version of how to debug the KUDO Kafka
with examples from plan status, service label mismatch and pod scheduling errors

This runbook will most likely be extended with more common use cases
**Which issue(s) this PR fixes**:

fixes one part of https://github.com/kudobuilder/kudo.dev/issues/151
**Special notes for your reviewer**:

### CODE REVIEW

File: content/docs/runbooks/kafka/debug-kafka.md [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+[2020-02-03 11:32:09,613] INFO [GroupMetadataManager brokerId=0] Removed 0 expired offsets in 0 milliseconds. (kafka.coordinator.group.GroupMetadataManager)
+[2020-02-03 11:42:09,613] INFO [GroupMetadataManager brokerId=0] Removed 0 expired offsets in 0 milliseconds. (kafka.coordinator.group.GroupMetadataManager)
+[2020-02-03 11:52:09,613] INFO [GroupMetadataManager brokerId=0] Removed 0 expired offsets in 0 milliseconds. (kafka.coordinator.group.GroupMetadataManager)
+[2020-02-03 12:02:09,613] INFO [GroupMetadataManager brokerId=0] Removed 0 expired offsets in 0 milliseconds. (kafka.coordinator.group.GroupMetadataManager)
+[2020-02-03 12:12:09,613] INFO [GroupMetadataManager brokerId=0] Removed 0 expired offsets in 0 milliseconds. (kafka.coordinator.group.GroupMetadataManager)
+[2020-02-03 12:22:09,613] INFO [GroupMetadataManager brokerId=0] Removed 0 expired offsets in 0 milliseconds. (kafka.coordinator.group.GroupMetadataManager)
+[2020-02-03 12:32:09,613] INFO [GroupMetadataManager brokerId=0] Removed 0 expired offsets in 0 milliseconds. (kafka.coordinator.group.GroupMetadataManager)
+```
+
+#### 4. Get logs from KUDO Kafka individual pod
```
Comment (meichstedt) [inline content/docs/runbooks/kafka/debug-kafka.md]:
```suggestion
#### 5. Get logs from KUDO Kafka individual pod
```

File: content/docs/runbooks/kafka/debug-kafka.md [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+
+`k logs kafka-instance-kafka-0 -c k8skafka`
```
Comment (meichstedt) [inline content/docs/runbooks/kafka/debug-kafka.md]:
```suggestion
`kubectl logs kafka-instance-kafka-0 -c k8skafka`
```

File: content/docs/runbooks/kafka/debug-kafka.md [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+[ ... lines removed for clarity ...]
+[2020-02-03 11:52:09,613] INFO [GroupMetadataManager brokerId=0] Removed 0 expired offsets in 0 milliseconds. (kafka.coordinator.group.GroupMetadataManager)
+[2020-02-03 12:02:09,613] INFO [GroupMetadataManager brokerId=0] Removed 0 expired offsets in 0 milliseconds. (kafka.coordinator.group.GroupMetadataManager)
+[2020-02-03 12:12:09,613] INFO [GroupMetadataManager brokerId=0] Removed 0 expired offsets in 0 milliseconds. (kafka.coordinator.group.GroupMetadataManager)
+[2020-02-03 12:22:09,613] INFO [GroupMetadataManager brokerId=0] Removed 0 expired offsets in 0 milliseconds. (kafka.coordinator.group.GroupMetadataManager)
+[2020-02-03 12:32:09,613] INFO [GroupMetadataManager brokerId=0] Removed 0 expired offsets in 0 milliseconds. (kafka.coordinator.group.GroupMetadataManager)
+[ ... lines removed for clarity ...]
+```
+
+#### 4. Get logs from KUDO Kafka node exporter container
```
Comment (meichstedt) [inline content/docs/runbooks/kafka/debug-kafka.md]:
```suggestion
#### 6. Get logs from KUDO Kafka node exporter container
```

File: content/docs/runbooks/kafka/debug-kafka.md [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+
+expected output is the logs from the node exporter container running as a sidecar with the broker 0 of Kafka:
+```
+time="2020-02-03T10:31:44Z" level=info msg="Starting node_exporter (version=0.18.1, branch=HEAD, revision=3db77732e925c08f675d7404a8c46466b2ece83e)" source="node_exporter.go:156"
+time="2020-02-03T10:31:44Z" level=info msg="Build context (go=go1.12.5, user=root@b50852a1acba, date=20190604-16:41:18)" source="node_exporter.go:157"
+```
+
+### Debugging service issues
+
+#### 5. Verify the service endpoints
```
Comment (meichstedt) [inline content/docs/runbooks/kafka/debug-kafka.md]:
```suggestion
#### 7. Verify the service endpoints
```

File: content/docs/runbooks/kafka/debug-kafka.md [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+`kubectl get endpoints kafka-instance-svc -o json | jq  -r '.subsets[].addresses[].hostname'`
+
+expected output is the pods name for the brokers:
+```
+kafka-instance-kafka-2
+kafka-instance-kafka-0
+kafka-instance-kafka-1
+```
+
+#### 6. Verify the service selector labels are matching the ones in pods
```
Comment (meichstedt) [inline content/docs/runbooks/kafka/debug-kafka.md]:
```suggestion
#### 8. Verify the service selector labels are matching the ones in pods
```

File: content/docs/runbooks/kafka/debug-kafka.md [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+```
+{
+  "app": "kafka",
+  "kudo.dev/instance": "kafka-instance"
+}
+```
+
+### Debugging health of all objects
+
+#### 7. Get list of all objects created by KUDO Kafka Instance
```
Comment (meichstedt) [inline content/docs/runbooks/kafka/debug-kafka.md]:
```suggestion
#### 9. Get list of all objects created by KUDO Kafka Instance
```

File: content/docs/runbooks/kafka/debug-kafka.md [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+poddisruptionbudget.policy/kafka-instance-pdb   N/A             1                 1                     5h29m
+NAME                                                           AGE
+rolebinding.rbac.authorization.k8s.io/kafka-instance-binding   5h29m
+NAME                                                 AGE
+role.rbac.authorization.k8s.io/kafka-instance-role   5h29m
+```
+
+### Debugging deployment issues
+
+#### 8. Pod stuck with Status `ContainerCreating`
```
Comment (meichstedt) [inline content/docs/runbooks/kafka/debug-kafka.md]:
```suggestion
#### 10. Pod stuck with Status `ContainerCreating` 
```

**Review by meichstedt — APPROVED** [review]
Looks good in general! few minor suggestions; I wonder whether the step numbers really help but I've at least tried to make it a sequence for now ;)

**Comment by zmalik** [comment]
thank you @meichstedt !
yeah steps could be an unordered list, but I didn't want to break from the runbook format. We can review that later and unordered steps if that makes sense for some runbooks like this one

**Pull request merged**